### DISCUSSION TITLE
HB_SCRIPT_DEVANAGARI not shaping properly

### DISCUSSION THREAD

**Discussion by anchor76** [harfbuzz/harfbuzz#2992] [Q&A] [answered]
hello,

i'm using harfbuzz for Arabic/Farsi, and Thai text shaping without problem. i started to shaping Hindi texts with HB_SCRIPT_DEVANAGARI script option, but the result seems incorrect for almost all words. anyone experienced problem with
Hindi text shaping? maybe i'm doing something wrong?
my harfbuz version is: 2.7.2

i made an example with the word: "OPTIONS" which should look: **विकल्प** 
[example image](https://fourflash.com/misc/harfbuzz.png)
on this image i showing the expected form, the non-shaped (stored) form, the HB_SCRIPT_DEVANAGARI result which is wrong, and the stored six unicode chars listed.

the very simple source code:
```
void cLocalization::doHindiShaping(const char* src, stringw& shaped)
{
	hb_buffer_t* buf;
	buf = hb_buffer_create();
	hb_buffer_add_utf8(buf, src, -1, 0, -1);
	hb_buffer_set_direction(buf, HB_DIRECTION_LTR);

	hb_buffer_set_script(buf, HB_SCRIPT_DEVANAGARI);

	hb_shape(hbFont, buf, NULL, 0);

	u32 glyph_count = 0;
	hb_glyph_info_t* glyph_info = hb_buffer_get_glyph_infos(buf, &glyph_count);
	stringw hindi;
	for (u32 _i = 0; _i < glyph_count; _i++)
	{
		hb_codepoint_t glyphid = glyph_info[_i].codepoint;
		wchar_t ch = font_glyph_lookup[glyphid];
		if (ch != wchar_t(0))
		{
			hindi.append(ch);
		}
	}

	hb_buffer_destroy(buf);

	shaped = hindi;
}

```

any help would be appreciated.

**Comment by khaledhosny** [comment, marked as answer]
What does `font_glyph_lookup` do here? You are converting glyph indices into `wchar_t` string which is doesn't sound right, how do you render the glyphs? You don't seem to be using glyph positioning information as well.

**Reply by anchor76** [reply]
thanks for your anaswer ;)

**Comment by behdad** [comment]
What Khaled said. If you want to convert shaping results back into a Unicode string, well, that doesn't work for most scripts, including Devanagari.

**Reply by anchor76** [reply]
thanks for your anaswer ;)
